lib/internal/modules/esm/translators.js

```diff
@@ -184,30 +254,58 @@ translators.set('commonjs', async function commonjsStrategy(url, source,
     }
     this.setExport('default', exports);
   });
+
+}
+
+translators.set('require-commonjs', (url, source, isMain) => {
+  assert(cjsParse);
+
+  return createCJSModuleWrap(url, source);
 });
 
-function cjsPreparseModuleExports(filename) {
+translators.set('commonjs', async function commonjsStrategy(url, source,
```

Comment:
> do we need two translators for CommonJS?

One is sync, the other one is async. Yes I think we need two of them.



> Is one `commonjs-via-commonjs-loader` and the other `commonjs-via-esm-loader`?

The CJS loader never calls into the ESM one, so all calls to those translator originates from the ESM one.

> Or is one `commonjs-via-require` and the other `commonjs-via-import`?

Yeah, well the async one covers more than just import (e.g. the entry point is also loaded async).

Reply:
How about `commonjs-sync` and `commonjs-async`?

Reply:
Well CJS is never async, that could be confusing. That being said, this is internal only, so I'm not going to fight for it, I'm happy to use whatever.

Reply:
I suggested those names because you wrote “One is sync, the other one is async.” Yet you also say that CJS is never async. So maybe you should explain in more detail the differences between these two translators and when each is used.

Reply:
One translator is async, the other one is sync. CJS modules are sync, in the sense they do not support TLA.

Reply:
> The other one is used for all the other cases when we want to load CJS.

Those other use cases are just `import` or the entry point, right? Assuming the entry is being handled by the ESM loader because of `--import` or `--loader` usage. So maybe just `require-commonjs` and `import-commonjs`, with explanatory comments:

```js
// Handle CommonJS modules referenced by `require` calls.
// This translator function must be sync, as `require` is sync.
translators.set('require-commonjs', (url, source, isMain) => {
  assert(cjsParse);

  return createCJSModuleWrap(url, source);
});

// Handle CommonJS modules referenced by `import` statements or expressions,
// or as the initial entry point when the ESM loader handles a CommonJS entry.
translators.set('import-commonjs', async function commonjsStrategy(url, source,
```